doc/sphinx/addendum/parallel-proof-processing.rst

```diff
@@ -55,6 +57,33 @@ proof does not begin with :cmd:`Proof using`, the system records in an
 auxiliary file, produced along with the ``.vo`` file, the list of section
 variables used.
 
+If a theorem has an incorrect annotation that omits a needed variable, you may see
+a message like this:
+
+   .. code-block::
+
+      File "./Pff.v", line 2372, characters 0-4:
+      Error: The following section variable is used but not declared:
+      precisionNotZero.
+
+      You can either update your proof to not depend on precisionNotZero, or you can
+      update your Proof line from
+      Proof using FtoRradix b pGivesBound precision radix radixMoreThanOne radixMoreThanZERO
+      to
+      Proof using FtoRradix b pGivesBound precision precisionNotZero radix radixMoreThanOne radixMoreThanZERO
+
+In this cse the minimal annotation suggested by the :flag:`Suggest Proof Using` flag is
+`Print Using pGivesBound precisionNotZero radixMoreThanOne.`  The other variables
+in the suggestion are unnecessary because they will be transitively included from
+the minimal annotation.
+
+Occasionally an incorrect annotation may produce a message in a theorem such as
+`(in proof FPredDiff1) Attempt to save an incomplete proof`.  Generally this indicates
+that an earlier theorem referenced by the theorem getting the error has an incorrect
+annotation.  The error will occur at the first reference to the earlier theorem.
```

Comment:
The following would be more accurate (but I'm not too happy with it yet):
```suggestion
Inferred annotations only mention needed variables. Annotations that mention unneeded variables will change the theorem type, and might break existing clients.
```

The reason is that errors might arise never/not at the first reference/have essentially arbitrary error messages.

Reply:
The text should include the message--that's what the user will see.  What you suggest is much less concrete.  "Inferred annotations" - how are they inferred?  How can you get this not at the first reference?  Can you give an example of breaking an existing client?

Reply:
OK, here's an example:

![image](https://user-images.githubusercontent.com/1253341/173270335-f3300f22-c7f2-49e0-905d-47847ca90f6c.png)

It's rather curious to have a proof take arguments and that the only way to set that up AFAIK is with `Proof using`.  Is that ever useful outside of managing async proofs?  It's unclear how the order of the arguments is determined.

Also, is `All` ever a useful option for `Proof using`?  How about `Type*` or `radixNotZero*`?  Isn't some automatic transitive closure applied even without the `*`?  The doc doesn't specify what relation the transitive closure is applied, either.

Not sure what, if any, of this is worth documenting.

Reply:
> The text should include the message--that's what the user will see.

I completely agree, but it should also not say the users things that are not true. What you wrote was generalized from one example. We cannot list all possible errors arising from "your theorem now takes 1 more argument than you expected". Especially, clients might not break when that argument is inferred via type classes.

> It's rather curious to have a proof take arguments and that the only way to set that up AFAIK is with Proof using.

Most theorems take arguments. Sections are just a way to state many theorems over the same arguments, and `Proof using` is only one way to set that up. Try writing statements about `q` and `radix`, and see their inferred types without `Proof.` Here's a made-up example:

```
Lemma square_equal : 0 < q <- q ^ 2 = radix^2 -> q = radix.
Proof. Admitted.
```

After the section, this will take two arguments (`q` and `radix`). BTW, this theorem is false without assuming `radixNotZero` — so `Proof using radixNotZero.` is more appropriate.

> "Inferred annotations" - how are they inferred?

Here's the current description of this inference process. We'll want more editing to make this connection more obvious.

> When a section ends, Coq looks at the proof object to decide which
section variables are actually used and hence have to be quantified in
the statement of the theorem.

Reply:
> I completely agree, but it should also not say the users things that are not true. What you wrote was generalized from one example. We cannot list all possible errors arising from "your theorem now takes 1 more argument than you expected". Especially, clients might not break when that argument is inferred via type classes.

I didn't generalize.  The wording is "an incorrect annotation **_may_** produce a message".  I don't think it's necessary to describe all the possible messages and why they appear.  I can mention other messages if you think they're important enough, but you'd need to provide the basic info on them--me finding them on my own by trial and error takes forever.

I guess the "The error will occur at the first reference to the earlier theorem." should go away?

> Most theorems take arguments.

Implicit, yes, but explicit arguments??

I got this error on your example:

![image](https://user-images.githubusercontent.com/1253341/173421740-2f33e770-d10c-466f-b5b1-eb7e60c30832.png)

You would then apply `square_equal` with a tactic like `apply (square_equal <ARGS>)`?

(Maybe off topic: a long while ago, I was trying to figure out how to use dependent types but even after searching the doc, internet and books over a couple days, I couldn't find the concrete syntax to make it work.  Perhaps this is what I needed?  I was trying to support arrays of length N and assure that subscripts were in bounds.  At some point I'd like to add a doc section on using dependent types in Coq.  But not today.)

> We'll want more editing to make this connection more obvious.

>> When a section ends, Coq looks at the proof object ...

How about adding this example: two variables defined in a section, a lemma references one and not the other, then `Print` the lemma and note that only one is included?

Reply:
Just fixed a typo: the thm was `Lemma square_equal : 0 < q -> q ^ 2 = radix^2 -> q = radix.` with `->` not `<-`.

>  I don't think it's necessary to describe all the possible messages and why they appear. I can mention other messages if you think they're important enough, but you'd need to provide the basic info on them--me finding them on my own by trial and error takes forever.

Listing a specific one would suggest that if an error message is completely different, it's unrelated — which is not misleading. I'd rather explain what changes ("a theorem takes unexpected arguments") so people can debug their errors.

Admittedly, I haven't demonstrated the unrelated examples; doing it properly would be some work...

> Implicit, yes, but explicit arguments??

Yep. For instance, `Nat.add_comm` takes two explicit arguments, and `Nat.add_le_mono` takes 6.

```coq
About Nat.add_comm.
About Nat.add_le_mono.
(*
Nat.add_comm : forall n m : nat, n + m = m + n

Nat.add_comm is not universe polymorphic
Arguments Nat.add_comm (n m)%nat_scope
Nat.add_comm is opaque
Expands to: Constant Coq.Arith.PeanoNat.Nat.add_comm
Nat.add_le_mono : forall n m p q : nat, n <= m -> p <= q -> n + p <= m + q

Nat.add_le_mono is not universe polymorphic
Arguments Nat.add_le_mono (n m p q)%nat_scope _ _
Nat.add_le_mono is opaque
Expands to: Constant Coq.Arith.PeanoNat.Nat.add_le_mono
*)
```

It's often hard to tell them apart, since many tactics (like `(e)apply` or `rewrite`) can infer both implicit and explicit arguments — while `Set Implicit Arguments` will silently make some arguments implicit.

> You would then apply square_equal with a tactic like apply (square_equal <ARGS>)?

That's one way; `apply square_equal`, `apply (square_equal <INITIAL_SEQUENCE_OF_ARGS>)`, or just using `square_equal <ARGS>` inside a Gallina term.